.changeset/spicy-trains-provide.md

```diff
@@ -0,0 +1,6 @@
+---
+"@fuel-ts/contract": minor
+"@fuel-ts/program": minor
+---
+
+feat!: implement non-blocking contract call
```

Comment:
The breaking changes in this one seem very problematic and will affect all our users without exception. Have you considered adding a new method, as in #2597?

Reply:
@arboleya I've modified the `call` method to be non-blocking. I believe this breaking change ensures that users are aware of the update and can adjust their projects accordingly. 

This modification reflects the natural behavior of a blockchain environment, where transaction processing can take several seconds, especially during network congestion. This approach is also consistent with how other popular web3 libraries handle transaction submissions.

To accommodate users who prefer a call that will resolve only after the transaction is processed, I've implemented a new method, `callAndWait`, which retains the previous behavior of the `call` method.

Using a flag within the same method to switch between blocking and non-blocking behaviors would result in different response types, potentially leading to unexpected results. Having two distinct methods avoids this issue and simplifies the interface, in my opinion.

After reflecting on this, I believe it makes sense to modify https://github.com/FuelLabs/fuels-ts/pull/2597 to mirror this approach. Where, `deployContract` would be a non-blocking call, and a new method, `deployAndWait`, would serve as the blocking call, resolving only when the contract is deployed.

Reply:
So the current playoff is what we consider the expected behaviour vs scale of the breaking change. Firstly, I do agree with separate methods. Given the impact, it makes sense to make the breakage as obvious as possible. 

Is there any other case for making the blocking call the default behaviour of `call`/`deployContract` other than it being the natural behaviour of a blockchain environment (which is a strong case)?

Because currently edging on the side of caution may make sense given current timescales and rollout. These methods could remain blocking and we have `callAsync`/`deployContractAsync` for non-blocking. Users that want this change can be made aware of it. 

Then in `1.0.0`, we could move to `call`/`deployContract` being blocking and `callSync`/`deployContractSync` (or `...andWait`). Given the major upgrade such a change would be expected.

Reply:
@danielbate I agree with your concerns, and they indeed make sense.

However, I see this as an opportunity to fix the behavior that was initially implemented using a non-ideal approach.

I believe the `BaseInvocationScope` and the `ContractFactory` should have been implemented with a similar logic to the TransactionResponse class, which from the beginning, was never a blocking call. Users need to call `waitForResult` when they want to wait until the transaction is processed.

Now that the main-net has not been released yet, this is the perfect time to make this breaking change, as it makes complete sense in the blockchain environment. This way, all users will be aware of it and can update their code accordingly.

Reply:
I guess it's now or`1.0.0`. I like the parity with #2597 so I will go with you on this one. Tricky to understand the impact this will have but I assume pretty significant. Let's make sure it's clearly communicated in the release notes. They may need some manual intervention.